File: ADT/dfs_bfs/797.py
# Directed Acyclic Graph(DAG)
# Amzaon, Microsoft

# DFS + Backtracking
# Tc = O(V+E)

class Solution:
    def dfs(
        self,
        graph: List[List[int]],
        u: int,
        target: int,
        result: List[List[int]],
        temp: List[int],
    ) -> None:
        temp.append(u)
        if u == target:
            result.append(temp[:])
        else:
            for v in graph[u]:
                self.dfs(graph, v, target, result, temp)
        temp.pop()

    def allPathsSourceTarget(self, graph: List[List[int]]) -> List[List[int]]:
        n = len(graph)
        source = 0
        target = n - 1
        result = []
        temp = []
        self.dfs(graph, source, target, result, temp)
        return result

File: ADT/dfs_bfs/797.js
/**
 * @param {number[][]} graph
 * @return {number[][]}
 */
var allPathsSourceTarget = function (graph) {
    const dfs = (start, target, result, temp) => {
        temp.push(start)
        if (start === target) {
            result.push([...temp])
        } else {
            for (const v of graph[start]) {
                dfs(v, target, result, temp)
            }
        }
        temp.pop()
    }
    const n = graph.length
    const source = 0
    const target = n - 1
    const result = []
    const temp = []
    dfs(source, target, result, temp)
    return result
};